# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.


configfile: "config/config.yaml"


read_type = config["read_type"]
sample_config = config["sample"]


bam_read = (
    "resources/example-reads-pacbio.bam"
    if read_type == "PacBio"
    else "resources/example-reads-illumina.bam"
)
sam_read = (
    "resources/example-reads-pacbio.sam"
    if read_type == "PacBio"
    else "resources/example-reads-illumina.sam"
)


rule get_genome:
    output:
        "resources/genome.fasta",
    params:
        species=sample_config["species"],
        datatype=sample_config["datatype"],
        build=sample_config["build"],
        release=sample_config["release"],
    log:
        "logs/get_genome.log",
    cache: "omit-software"  # save space and time with between workflow caching (see docs)
    wrapper:
        "v2.3.2/bio/reference/ensembl-sequence"


rule get_chromosome:
    output:
        "resources/chr1.fasta",
    params:
        species=sample_config["species"],
        datatype=sample_config["datatype"],
        build=sample_config["build"],
        release=sample_config["release"],
        chromosome=sample_config["chromosome"],
    log:
        "logs/get_chromosome.log",
    cache: "omit-software"  # save space and time with between workflow caching (see docs)
    wrapper:
        "v2.3.2/bio/reference/ensembl-sequence"


rule find_candidates:
    input:
        "resources/example-genome.fasta",
    output:
        "resources/example-candidates.bcf",
    log:
        "logs/find_candidates.log",
    conda:
        "envs/rust.yaml"
    params:
        varlo_path=config["varlo_path"],
        pipeline_path=config["pipeline_path"],
    shell:
        """ 
        cd {params.varlo_path}
        cargo run -- methylation-candidates {params.pipeline_path}{input} {params.pipeline_path}/{output}
        """


rule sam_to_bam:
    input:
        sam_read,
    output:
        bam_read,
    log:
        "logs/sam_to_bam.log",
    conda:
        "envs/samtools.yaml"
    params:
        pipeline_path=config["pipeline_path"],
    shell:
        "samtools view -bS {params.pipeline_path}{input} > {params.pipeline_path}/{output}"


rule compute_meth_observations:
    input:
        fasta="resources/example-genome.fasta",
        bam=bam_read,
        candidates="resources/example-candidates.bcf",
    output:
        "resources/normal.bcf",
    log:
        "logs/copute.log",
    conda:
        "envs/rust.yaml"
    params:
        varlo_path=config["varlo_path"],
        pipeline_path=config["pipeline_path"],
        read_type=config["read_type"],
    shell:
        """ 
        cd {params.varlo_path}
        cargo run -- preprocess variants {params.pipeline_path}{input.fasta} --candidates {params.pipeline_path}{input.candidates} --bam {params.pipeline_path}{input.bam} --read-type {params.read_type}> {params.pipeline_path}{output}
        """


rule observations_to_vcf:
    input:
        "resources/normal.bcf",
    output:
        "resources/normal.vcf",
    conda:
        "envs/samtools.yaml"
    log:
        "logs/convert_to_vcf.log",
    shell:
        """
        bcftools view {input} > {output}
        """

rule call_methylation:
    input:
        preprocess_obs="resources/normal.bcf",
        scenario="resources/scenario.yaml"
    output:
        "resources/calls.bcf",
    log:
        "logs/copute.log",
    conda:
        "envs/rust.yaml"
    params:
        varlo_path=config["varlo_path"],
        pipeline_path=config["pipeline_path"],
        read_type=config["read_type"],
    shell:
        """ 
        cd {params.varlo_path}
        cargo run -- call variants generic  --scenario {params.pipeline_path}{input.scenario} --obs normal={params.pipeline_path}{input.preprocess_obs} > {params.pipeline_path}/{output}
        """

rule calls_to_vcf:
    input:
        "resources/calls.bcf",
    output:
        "resources/calls.vcf",
    conda:
        "envs/samtools.yaml"
    log:
        "logs/convert_to_vcf.log",
    shell:
        """
        bcftools view {input} > {output}
        """